src/SignalR/common/Shared/MessageBuffer.cs

```diff
@@ -0,0 +1,512 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR.Internal;
+
+internal sealed class MessageBuffer : IDisposable
+{
+    private static readonly TaskCompletionSource<FlushResult> _completedTCS = new TaskCompletionSource<FlushResult>();
+
+    private readonly ConnectionContext _connection;
+    private readonly IHubProtocol _protocol;
+    private readonly AckMessage _ackMessage = new(0);
+    private readonly SequenceMessage _sequenceMessage = new(0);
+    private readonly Channel<int> _waitForAck = Channel.CreateBounded<int>(new BoundedChannelOptions(1) { FullMode = BoundedChannelFullMode.DropOldest });
+    private readonly int _bufferLimit = 100 * 1000;
+
+#if NET8_0_OR_GREATER
+    private readonly PeriodicTimer _timer = new(TimeSpan.FromSeconds(1));
+#else
+    private readonly TimerAwaitable _timer = new(TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
+#endif
+    private readonly SemaphoreSlim _writeLock = new(1, 1);
+
+    private long _totalMessageCount;
+    private bool _waitForSequenceMessage;
+
+    // Message IDs start at 1 and always increment by 1
+    private long _currentReceivingSequenceId = 1;
+    private long _latestReceivedSequenceId = long.MinValue;
+    private long _lastAckedId = long.MinValue;
+
+    private TaskCompletionSource<FlushResult> _resend = _completedTCS;
+
+    private object Lock => _buffer;
+
+    private LinkedBuffer _buffer;
+    private int _bufferedByteCount;
+
```

Comment:
Consider, for the preview release, making this cheaply configurable (e.g. via environment variable) or at least just making it much larger.